### PR TITLE
add mask as channel in preprocessing

### PR DESCRIPTION
**Pull request by Soorya19Pradeep**
Changes in preprocessing in the branch:
1. Removed flatfielding in preprocessing.
2. Dataset normalization values stored in plate zattrs file.
3. FOV statistics stored in position zattrs.
4. 3D segmentation transferd to branch.

### CODE REVIEW

**Comment by mattersoflight** [comment]
Thanks @Soorya19Pradeep. Looks good to me. Since @ziw-liu is working on `rm-flatfield-from-train` branch, please merge after his review.

One question I couldn't get an answer to while scanning the code:
Does the gunpowder data loader load the masks that match with a specific channel based on the metadata stored in `.zattrs` file OR does the data loader assume that the mask channel has a specific name (e.g., `otsu_mask`)? 

At this point, it is sufficient that we assign mask channels the match that links them with the channel (e.g., `nuc_mask`, `membrane_mask`). We can refine the data-loading/augmentation strategy when we meet next offline.

**Comment by Soorya19Pradeep** [comment]
Hello @mattersoflight. Thank you for looking into the PR. 

The current version of the code names the mask channels based on the channel name used for segmentation, adding '_mask' at the the end of the channel name. For instance, if the channel to be segmented is 'Deconvolved-Nucl'. The mask channel name will be 'Deconvolved-Nucl_mask'. I have tested the chnages on this dataset if you want to see the the results.
`/local/scratch/groups/cmanalysis.grp/microDL_SP/Input/all_pos_Phase5e-4_Denconv_Nuc8e-4_Mem8e-4_pad15_bg50_registeredNGFF_32slices.zarr`

**Comment by mattersoflight** [comment]
> The mask channel name will be 'Deconvolved-Nucl_mask'. I have tested the changes on this dataset if you want to see the results.
👍🏼 After browsing the datasets, I realized that masks are currently written as `float32` arrays. They should be written as `bool` and assigned metadata such that they appear as a label layer in viewers.

**Comment by mattersoflight** [comment]
Another suggestion: if the 3D segmentation code is now part of the `iohub_dependency` branch, please delete the `preprocessing_3DSegmentation` branch. Looking at the commit graph, the commit history looks quite linear between other 3 active branches.

**Comment by Soorya19Pradeep** [comment]
> > The mask channel name will be 'Deconvolved-Nucl_mask'. I have tested the changes on this dataset if you want to see the results.
> 
> 👍🏼 After browsing the datasets, I realized that masks are currently written as `float32` arrays. They should be written as `bool` and assigned metadata such that they appear as napari label layers (https://napari.org/stable/howtos/layers/labels.html).

On discussion we decided to store the mask as the same datatype as the segmented channel. On loading the mask as label layer to napari, the features in napari related to visualization of label layers not flexible to use. So it doesn't add an advantage for visualization.

**Comment by Soorya19Pradeep** [comment]
@ziw-liu, @mattersoflight, I have added a documentation markdown file explaining the changes on dataset during preprocessing. Please let me know if I missed out something and I will add that in.

File: micro_dl/preprocessing/preprocessing.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+* Segments the target images using selected segmentation algorithm in the configuration file(otsu, mem_detection, unimodal).
+
+* Stores the mask output as an extra channel in the zarr store with the name of the segmented channel with added subscript '_mask'. For instance, if the user segments the channel named 'Deconvolved-Nuc', the mask channel added will be called 'Deconvolved-Nuc_mask'. The datatype is the same as the channel which is segmented (most possibly float32).
+
+* Stores the information related to normalization of all input channels mentioned in the configuration. The dataset statistics are stored in the plate level .zattrs file, while the information specific to the position is added in the .zattrs file at each position. The details are explained below.
+
+Here is the structure of a 0.4 NGFF version zarr store wriiten using [iohub](https://github.com/czbiohub/iohub) for a dataset with a single condition and multiple imaging FOVs.
+
+```
+well0
```
Comment (ziw-liu) [inline micro_dl/preprocessing/preprocessing.md:14]: I'm not sure if it's a good idea to name this example `well0` since it's really a well but a row in NGFF.
Also it would be clearer to include the top level `.zarr` directory in the tree like [this](https://ngff.openmicroscopy.org/0.4/index.html#hcs-layout).

Comment (Soorya19Pradeep) [reply micro_dl/preprocessing/preprocessing.md]: Thank you @ziw-liu . This is a much clearer example. I have made the modification accordingly.

File: micro_dl/utils/io_utils.py [hunk 2]
```diff
@@ -46,9 +46,21 @@ def write_meta_field(position: ngff.Position, metadata, field_name):
...
     if field_name in position.zattrs:
-        position.zattrs[field_name].update(metadata)
+        if subfield_name in position.zattrs[field_name]:
+            position.zattrs[field_name][subfield_name].update(metadata)
+        else:
+            D1 = position.zattrs[field_name]
+            field_metadata = {
+                subfield_name: metadata,
+            }
+            # position.zattrs[field_name][subfield_name] = metadata
```
Comment (ziw-liu) [inline micro_dl/utils/io_utils.py:56]: I'm not sure if I read this correctly: is this avoiding overwriting existing `subfield_name` and attempting to duplicate the key?

Comment (ziw-liu) [reply micro_dl/utils/io_utils.py:56]: Also if at one level there will always be only one `subfield_name` why is it necessary?

File: micro_dl/utils/mp_utils.py [hunk 3]
```diff
@@ -141,38 +141,14 @@ def create_and_write_mask(
 
     # save masks as additional channel
     position_masks = position_masks.astype(position.data.dtype)
+    new_channel_name = channel_name + '_mask'
     io_utils.add_channel(
```
Comment (ziw-liu) [inline micro_dl/utils/mp_utils.py:145]: This function is also only called in this file. Maybe we should remove `io_utils.py` altogether since we don't have 'untracked' arrays anymore.

File: micro_dl/utils/meta_utils.py [hunk 4]
```diff
@@ -53,15 +53,16 @@ def generate_normalization_metadata(
...
 
     # sample values and use them to get normalization statistics
-    for i, channel in enumerate(channel_ids):
+    for i,channel in enumerate(channel_ids):
         show_progress_bar(
             dataloader=channel_ids,
             current=i,
             process="sampling channel values",
         )
+        print(channel)
```
Comment (ziw-liu) [inline micro_dl/utils/meta_utils.py:65]: Is this intended?

File: micro_dl/utils/meta_utils.py [hunk 5]
```diff
@@ -53,15 +53,16 @@ def generate_normalization_metadata(
     # get arguments for multiprocessed grid sampling
     mp_grid_sampler_args = []
     for _, position in position_map:
-        mp_grid_sampler_args.append([position, True, grid_spacing])
+        mp_grid_sampler_args.append([position, grid_spacing])
 
     # sample values and use them to get normalization statistics
-    for i, channel in enumerate(channel_ids):
+    for i,channel in enumerate(channel_ids):
```
Comment (ziw-liu) [inline micro_dl/utils/meta_utils.py:59]: Accidental edit?

File: micro_dl/utils/io_utils.py (hunk 2 above)
Comment (Soorya19Pradeep) [reply micro_dl/utils/io_utils.py]: I have deleted the `io_utils.py` file and transferred the two functions to `meta_utils.py` file. The functions are also called in `mp_utils.py` file, so I have changed it there as well.

I have edited the docstring as well.

File: micro_dl/utils/meta_utils.py (hunk 5 above)
Comment (Soorya19Pradeep) [reply micro_dl/utils/meta_utils.py]: Yes. Removed. Thanks!

File: micro_dl/utils/meta_utils.py (hunk 4 above)
Comment (Soorya19Pradeep) [reply micro_dl/utils/meta_utils.py]: No. Accidentally left during testing. Removed. Thanks!

**Review by mattersoflight — APPROVED** [review]
@Soorya19Pradeep looks good to me. The only check that fails right now is [linting](https://code.visualstudio.com/docs/python/linting). See the Details in the build log. 

Please merge after you have fixed this.

File: micro_dl/preprocessing/preprocessing.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+* Segments the target images using selected segmentation algorithm in the configuration file(otsu, mem_detection, unimodal).
+
+* Stores the mask output as an extra channel in the zarr store with the name of the segmented channel with added subscript '_mask'. For instance, if the user segments the channel named 'Deconvolved-Nuc', the mask channel added will be called 'Deconvolved-Nuc_mask'. The datatype is the same as the channel which is segmented (most possibly float32).
+
+* Stores the information related to normalization of all input channels mentioned in the configuration. The dataset statistics are stored in the plate level .zattrs file, while the information specific to the position is added in the .zattrs file at each position. The details are explained below.
+
+Here is the structure of a 0.4 NGFF version HCS format zarr store wriiten using [iohub](https://github.com/czbiohub/iohub) for a dataset with a single condition and multiple imaging FOVs.
+
+```buildoutcfg
```
Comment (ziw-liu) [inline micro_dl/preprocessing/preprocessing.md:13]: This language seems irrelevant, maybe use `text` instead?

File: micro_dl/preprocessing/preprocessing.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    │   │   │   ├── 0         # (T, C, Z, Y, X) float32
+    │   │   │   │   ...       # Resolution levels
+    │   │   │   ├── n
+    │   │   │   └── labels    # Labels (optional)
+
+ ```
+
+The statistics are added as dictionaries into the .zattrs file. An example of plate level metadata is here:
+
+```
```
Comment (ziw-liu) [inline micro_dl/preprocessing/preprocessing.md:69]: These two blocks can be annotated as `json`.

File: micro_dl/utils/meta_utils.py [hunk 8]
```diff
@@ -3,15 +3,89 @@
...
+            # position.zattrs[field_name][subfield_name] = metadata
+            position.zattrs[field_name] = {**D1, **field_metadata}
+    else:
+        field_metadata = {
+            subfield_name: metadata,
+        }
+        position.zattrs[field_name] = field_metadata
+
+
+def add_channel(
```
Comment (ziw-liu) [inline micro_dl/utils/meta_utils.py:45]: This function is only used once in `mp_utils.py`. I think it can be moved there instead.

Comment (Soorya19Pradeep) [reply micro_dl/utils/meta_utils.py]: Makes sense! I have moved it to `mp_utils.py`. Thanks!

File: micro_dl/preprocessing/preprocessing.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    │   │   │   ├── 0         # (T, C, Z, Y, X) float32
+    │   │   │   │   ...       # Resolution levels
+    │   │   │   ├── n
+    │   │   │   └── labels    # Labels (optional)
+    |   |
+    |   ├── 001               # second FOV
+
+ ```
+
+ Here the dataset statistics is stored inside the 'plate' folder and the position statistics is stored in '.zattrs' inside plate/A/1/0 folder.
```
Comment (ziw-liu) [inline micro_dl/preprocessing/preprocessing.md:39]: Extra 1-space indentation?

File: micro_dl/preprocessing/preprocessing.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+* Segments the target images using selected segmentation algorithm in the configuration file(otsu, mem_detection, unimodal).
+
+* Stores the mask output as an extra channel in the zarr store with the name of the segmented channel with added subscript '_mask'. For instance, if the user segments the channel named 'Deconvolved-Nuc', the mask channel added will be called 'Deconvolved-Nuc_mask'. The datatype is the same as the channel which is segmented (most possibly float32).
+
+* Stores the information related to normalization of all input channels mentioned in the configuration. The dataset statistics are stored in the plate level .zattrs file, while the information specific to the position is added in the .zattrs file at each position. The details are explained below.
+
+Here is the structure of a 0.4 NGFF version HCS format zarr store wriiten using [iohub](https://github.com/czbiohub/iohub) for a dataset with a single condition and multiple imaging FOVs.
+
+```text
+.                             # Root folder, potentially in S3,
```
Comment (ziw-liu) [inline micro_dl/preprocessing/preprocessing.md:14]: Probably want to remove/modify words like 'potentially in S3' and 'id=5966'. Also please cite the specification section from which these are copied with a link.
